File: back/src/controllers/CampaignController.ts
import { Request, Response } from "express";
import {CampaignService} from "../services/CampaignService";

const campaignService = new CampaignService();

export class CampaignController {

    static async createCampaign(req: Request, res : Response) {
        try {
            const { name, campaignId } = req.body;
            if (!name || !campaignId) {
                return res.status(400).json({ message: "Name and campaignId are required" });
            }
            const newCampaign = await campaignService.createCampaign({ name, campaignId });
            res.status(201).json(newCampaign);
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async getCampaignById(req: Request, res : Response) {
        try {
            const id = req.params.id;
            const campaign = await campaignService.getById(id);
            if (campaign) {
                res.status(200).json(campaign);
            } else {
                res.status(404).json({ message: "Campaign not found" });
            }
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async updateCampaign(req: Request, res : Response) {
        try {
            const campaignId = req.params.id;
            const campaign = await campaignService.getById(campaignId);
            if (!campaign) {
                return res.status(404).json({ message: "Campaign not found" });
            }
            const updatedCampaign = await campaignService.updateCampaign(campaign, req.body);
            if (updatedCampaign) {
                res.status(200).json(updatedCampaign);
            } else {
                res.status(404).json({ message: "Campaign not found" });
            }
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async deleteCampaign(req: Request, res : Response) {
        try {
            const campaignId = req.params.id;
            await campaignService.deleteCampaign(campaignId);
            res.status(204).send();
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async addUserToCampaign(req: Request, res : Response) {
        try {
            const { campaignId, userId } = req.body;
            const updatedCampaign = await campaignService.addUserToCampaign(campaignId, userId);
            res.status(200).json(updatedCampaign);
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async removeUserFromCampaign(req: Request, res : Response) {
        try {
            const { campaignId, userId } = req.body;
            const updatedCampaign = await campaignService.removeUserFromCampaign(campaignId, userId);
            res.status(200).json(updatedCampaign);
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }

    static async getCampaignByCampaignId(req: Request, res : Response) {
        try {
            const campaignId = req.params.campaignId;
            const campaign = await campaignService.getByCampaignId(campaignId);
            if (campaign) {
                res.status(200).json(campaign);
            } else {
                res.status(404).json({ message: "Campaign not found" });
            }
        } catch (error: any ) {
            res.status(500).json({ error: error.message });
        }
    }

    static async getAllCampaigns(req: Request, res : Response) {
        try {
            const campaigns = await campaignService.getAllCampaign();
            res.status(200).json(campaigns);
        } catch (error : any) {
            res.status(500).json({ error: error.message });
        }
    }
}

File: back/src/main/kotlin/com/apigateway/controller/CampaignController.kt
package com.apigateway.controller

import com.apigateway.dto.*
import com.apigateway.service.CampaignService
import jakarta.validation.Valid
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import java.util.UUID

@RestController
@RequestMapping("/campaign")
class CampaignController(
    private val campaignService: CampaignService
) {

    @PostMapping
    fun createCampaign(@Valid @RequestBody request: CreateCampaignRequest): ResponseEntity<CampaignResponse> {
        val campaign = campaignService.createCampaign(request)
        return ResponseEntity.status(HttpStatus.CREATED).body(campaign)
    }

    @GetMapping
    fun getAllCampaigns(): ResponseEntity<List<CampaignResponse>> {
        val campaigns = campaignService.getAllCampaigns()
        return ResponseEntity.ok(campaigns)
    }

    @GetMapping("/{id}")
    fun getCampaignById(@PathVariable id: UUID): ResponseEntity<CampaignResponse> {
        val campaign = campaignService.getById(id)
        return ResponseEntity.ok(campaign)
    }

    @GetMapping("/byCampain/{campaignId}")
    fun getCampaignByCampaignId(@PathVariable campaignId: String): ResponseEntity<CampaignResponse> {
        val campaign = campaignService.getByCampaignId(campaignId)
        return ResponseEntity.ok(campaign)
    }

    @PutMapping("/{id}")
    fun updateCampaign(
        @PathVariable id: String,
        @Valid @RequestBody request: UpdateCampaignRequest
    ): ResponseEntity<CampaignResponse> {
        val campaign = campaignService.updateCampaign(id, request)
        return ResponseEntity.ok(campaign)
    }

    @DeleteMapping("/{id}")
    fun deleteCampaign(@PathVariable id: UUID): ResponseEntity<Unit> {
        campaignService.deleteCampaign(id)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/{id}/results")
    fun getCampaignResults(@PathVariable id: UUID): ResponseEntity<Any> {
        val results = campaignService.getCampaignResults(id)
        return ResponseEntity.ok(results)
    }
}